### PR TITLE
3636 gist

### PR DESCRIPTION
**Pull request by luancazarine**
Triggers
  - New Contact (Polling)

Actions
  - Add Tag to an Existing Contact
  - Create or Update Contact
  - Remove Tag From Existing Contact
  - Find a Contact
  
 

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/3940"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/gist/gist.app.mjs [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+          name, id,
+        }) => ({
+          label: name,
+          value: JSON.stringify({
+            id,
+            name,
+          }),
+        })) || [];
+      },
+    },
```
Comment (michelle0927) [inline components/gist/gist.app.mjs:51]: This seems like a good opportunity to use `withLabel`. If you set `withLabel` to true, the prop value will contain `tagId.value` and `tagId.label`
```suggestion
    tagId: {
      label: "Tag",
      description: "The tag that will be added",
      type: "integer",
      async options({ page }) {
        const { tags } = await this.listTags({
          params: {
            page: page + 1,
          },
        });

        return tags.map(({
          name, id,
        }) => ({
          label: name,
          value: id,
        })) || [];
      },
      withLabel: true,
    },
```

File: components/gist/actions/add-tag-to-contact/add-tag-to-contact.mjs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    const data = {
+      ...tag,
+      contacts: this.contactId.map((contactId) => {
+        return {
+          id: `${contactId}`,
+        };
+      }),
+    };
```
Comment (michelle0927) [inline components/gist/actions/add-tag-to-contact/add-tag-to-contact.mjs:44]: If we change `tagId` to use `withLabel`, we can update this:
```suggestion
    tagId: {
      propDefinition: [
        gist,
        "tagId",
      ],
    },
  },
  async run({ $ }) {
    const data = {
      id: this.tagId.value,
      name: this.tagId.label,
      contacts: this.contactId.map((contactId) => {
        return {
          id: `${contactId}`,
        };
      }),
    }
```

File: components/gist/actions/add-tag-to-contact/add-tag-to-contact.mjs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    const response = await this.gist.updateTagToContact({
+      $,
+      data,
+    });
+
+    $.export("$summary", `Successfully added tag ${tag.name} to contact`);
```
Comment (michelle0927) [inline components/gist/actions/add-tag-to-contact/add-tag-to-contact.mjs:51]:
```suggestion
    $.export("$summary", `Successfully added tag ${this.tagId.label} to contact`);
```

File: components/gist/actions/remove-tag-from-contact/remove-tag-from-contact.mjs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  type: "action",
+  version: "0.0.1",
+  props: {
+    gist,
+    tagId: {
+      propDefinition: [
+        gist,
+        "tagId",
+      ],
+      type: "string",
```
Comment (michelle0927) [inline components/gist/actions/remove-tag-from-contact/remove-tag-from-contact.mjs:17]:
```suggestion
```

File: components/gist/actions/remove-tag-from-contact/remove-tag-from-contact.mjs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        };
+      }),
+    };
+
+    const response = await this.gist.updateTagToContact({
+      $,
+      data,
+    });
+
+    $.export("$summary", `Successfully removed tag ${tag.name} from contact${
```
Comment (michelle0927) [inline components/gist/actions/remove-tag-from-contact/remove-tag-from-contact.mjs:47]: If we change tagId to use withLabel, we can update this:
```suggestion
    const data = {
      id: this.tagId.value,
      name: this.tagId.label,
      contacts: this.contactId.map((contactId) => {
        return {
          id: `${contactId}`,
          untag: true,
        };
      }),
    };

    const response = await this.gist.updateTagToContact({
      $,
      data,
    });

    $.export("$summary", `Successfully removed tag ${this.tagId.label} from contact${
```

File: components/gist/actions/remove-tag-from-contact/remove-tag-from-contact.mjs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      this.contactId.lenght
+        ? "s"
+        : ""
+    }`);
```
Comment (michelle0927) [inline components/gist/actions/remove-tag-from-contact/remove-tag-from-contact.mjs:51]:
```suggestion
      this.contactId.length > 1
        ? "s"
        : ""
    }`);
```

**Review by michelle0927 — DISMISSED** [review]
Left some comments about updating the `tagId` prop to use `withLabel`. Note: I haven't tested anything, so make sure to test any changes.

File: components/gist/sources/new-contact/new-contact.mjs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+      const ts = Date.parse(event.created_at);
+      this.$emit(event, {
+        id: `${event.id}_${ts}`,
+        ts,
+        summary: `New contact created: ${event.full_name || event.email} (${event.id})`,
+      });
+    },
+  },
+  async run() {
+    console.log("Raw received event:");
```
Comment (andrewjschuang) [inline components/gist/sources/new-contact/new-contact.mjs:71]: I think the message may be misleading

```suggestion
    console.log("Run started:");
```

File: components/gist/actions/add-tag-to-contact/add-tag-to-contact.mjs [hunk 8]
```diff
@@ -0,0 +1,46 @@
+import gist from "../../gist.app.mjs";
+
+export default {
+  ...gist,
+  key: "gist-add-tag-to-contact",
```
Comment (andrewjschuang) [inline components/gist/actions/add-tag-to-contact/add-tag-to-contact.mjs:5]: Shouldn't this all be in plural (contacts)?

**Review by andrewjschuang — COMMENTED** [review]
I've left some comments below.
Oh, and `@pipedream/platform` dependency is missing in `package.json`

File: components/gist/actions/remove-tag-from-contact/remove-tag-from-contact.mjs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        }),
+      ],
+      type: "integer[]",
+    },
+  },
+  async run({ $ }) {
+    const data = {
+      id: this.tagId.value,
+      name: this.tagId.label,
+      contacts: this.contactId.map((contactId) => {
```
Comment (vellames) [inline components/gist/actions/remove-tag-from-contact/remove-tag-from-contact.mjs:33]:
```suggestion
      contacts: this.contactId.map((contactId) => ({
          id: String(contactId),
          untag: true,
      ))
```

**Review by vellames — CHANGES_REQUESTED** [review]
Hey, just left one comment

**Review by vellames — APPROVED** [review]
lgtm

**Comment by vunguyenhung** [comment]
Hello everyone, I have tested this PR and there're some test cases failed or needed improvement. 

Please check the test report below for more information
[Gist_3940_1550.pdf](https://github.com/PipedreamHQ/pipedream/files/9358005/Gist_3940_1550.pdf)

**Review by vellames — APPROVED** [review]

**Comment by luancazarine** [comment]
Waiting for resolution
https://github.com/PipedreamHQ/pipedream/issues/4021

**Review by michelle0927 — DISMISSED** [review]
LGTM!

**Review by michelle0927 — APPROVED** [review]

**Pull request merged**